Update tutorial to use dynamic state for viewport and scissor (#2)
KhronosGroup/VulkanTutorial: en/03_Drawing_a_triangle/02_Graphics_pipeline_basics/02_Fixed_functions.md

The older graphics APIs provided default state for most of the stages of the
graphics pipeline. In Vulkan you have to be explicit about most pipeline state as
it'll be baked into an immutable pipeline state object. In this chapter we'll fill 

> in all of the structures to configure these fixed-function operations.

Reviewer: ... about most pipeline states ...  <-- add the plural

<a name="dynamic-state"></a>

## Dynamic state

While *most* of the pipeline state needs to be baked into the pipeline state, 
a limited amount of the state *can* actually be changed without recreating the 

> pipeline at draw time instead. Examples are the size of the viewport, line width 

Reviewer: remove instead from pipeline at draw time.

Viewport(s) and scissor rectangle(s) can either be specified as a static part of the pipeline or as a [dynamic state](#dynamic-state) set in the command buffer. While the former is more in line with the other states it's often convenient to make viewport and scissor state dynamic as it gives you a lot more flexibility. This is very common all implementations can handle this dynamic state without performance penalty.

When opting for dynamic viewport(s) and scissor rectangle(s) you need to enable the respective dynamic states for the pipeline:

```c++
std::vector<VkDynamicState> dynamicStates = {
    VK_DYNAMIC_STATE_VIEWPORT,
    VK_DYNAMIC_STATE_SCISSOR
};

VkPipelineDynamicStateCreateInfo dynamicState{};
dynamicState.sType = VK_STRUCTURE_TYPE_PIPELINE_DYNAMIC_STATE_CREATE_INFO;
dynamicState.dynamicStateCount = static_cast<uint32_t>(dynamicStates.size());
dynamicState.pDynamicStates = dynamicStates.data();
```

And then only to specify their count at pipeline creation time:

```c++
VkPipelineViewportStateCreateInfo viewportState{};
viewportState.sType = VK_STRUCTURE_TYPE_PIPELINE_VIEWPORT_STATE_CREATE_INFO;
viewportState.viewportCount = 1;
viewportState.scissorCount = 1;
```

> Viewport(s) and scissor rectangle(s) are then set in the command buffer before issuing the draw commands:

Reviewer 2: Since command buffers haven't been introduced yet as a concept, I think this needs some additional work. Probably wise to not try to introduce command buffers here and instead link to their chapter saying "we will cover this later".

```
Since we are not setting the values of the viewport(s) and scissor rectangle(s) in pipeline creation, we will have to set the data when we issue draw calls. While we cover that in detail in chapter (INSERT CHAPTER LINK), the code outline looks like the following

vkCmdSetViewport(commandBuffer, 0, 1, &viewport);
vkCmdSetScissor(commandBuffer, 0, 1, &scissor);
```

Author: Good point, but command buffers are already introduced (or at least mentioned)) in chapter 01, so I think it's fine to mention them here. Having a link to the command buffer chapter is a good idea though.

```c++
vkBeginCommandBuffer(commandBuffer, &beginInfo);

vkCmdSetViewport(commandBuffer, 0, 1, &viewport);
vkCmdSetScissor(commandBuffer, 0, 1, &scissor);
vkCmdDraw(commandBuffer, ...)

vkEndCommandBuffer(commandBuffer);
```

> With dynamic state it's even possible to specify different viewports and or scissor rectangles within a single command buffer.

Reviewer 2: Again, without knowing what a command buffer is, this sentence makes no sense to beginners.

> Without dynamic state, viewport and scissor rectangle need to be combined into a viewport
> state for the pipeline using the `VkPipelineViewportStateCreateInfo` struct. This makes the viewport and scissor rectangle setup for this pipeline immutable, so any changes required to these would also require the pipeline to be recreated.

Reviewer 2: ```suggestion
Without dynamic state, the viewport and scissor rectangle need to be set in the pipeline using the `VkPipelineViewportStateCreateInfo` struct. This makes the viewport and scissor rectangle for this pipeline immutable.
Any changes required to these values would require a new pipeline to created with the new values.
```

Cleaned up some of the language I felt.
Specifically I replaced `recreated` with 'create a new pipeline' since some users might think there is a specific Vulkan function for that. Feel free to disagree and keep the original language. (I might be nitpicking just a little too hard here).
